### ISSUE TITLE
REPROG_CONTROLS_V2 for Harmony Companion remote

### ISSUE THREAD

**Issue by jontywigster** [pwr-Solaar/Solaar#1174]
Hello, 

Back in [#1019], @pfps mentioned -

> This device has a useful feature that Solaar does not support - REPROG_CONTROLS_V2. If you are interested in changing how keys work and can test new version of Solaar it may be possible to add this capability to Solaar.

Where could I see what REPROG_CONTROLS_V2 would do?

The reason I ask is I'm clutching at straws. I have a Harmony Companion remote, paired to a Logitech Unifying Receiver. Various buttons (navigation and numbers) appear to do nothing, i.e. usbmon shows no traffic. The buttons work if paired with a Harmony Hub. I wonder if some sort of setting might make pressing these buttons visible to Linux, whether a Solaar rule or anything else. I'm currently running Solaar 1.0.6, the remote is shown as 'Harmony 20+', no settings are available to change.

Thanks and sorry for being a pest.

**Comment by pfps** [comment]
I'm not sure what can be done.  I don't have documentation for these features, but I'm asking around for it.
It might be the case that the feature is very similar to REPROG_CONTROLS_V4.  This would allow changing the action associated with a key.
But the first thing to do is to get the output from `solaar show` to see what Solaar knows about the device already.   Then do `solaar -dd show` and paste this output, which might have some clues as to what can be done.

**Comment by jontywigster** [comment]
Hello, 

Thanks very much indeed for getting back to me. 

I've attached the output of both files -

[solaar_show.txt](https://github.com/pwr-Solaar/Solaar/files/6487370/solaar_show.txt)
[solaar_dd_show.txt](https://github.com/pwr-Solaar/Solaar/files/6487371/solaar_dd_show.txt)

The additional output from -dd is polluted with lots of button presses on the remote. I thought it might be useful to get to the end of listing features but the remote goes offline pretty quickly unless a button is pressed to keep it awake. Apologies if this wasn't good.

Thanks again.

**Comment by pfps** [comment]
The thing I was looking for was the interaction between Solaar and the device when Solaar is determining which reprogrammable keys are available.  However, there is a regression in Solaar that uses the oldest version of the feature instead of the second version.  Your device has both.

I put together a PR #1175 to fix this regression.  Please clone Solaar, download this PR - directions are in the PR, and try `bin/solaar -dd show` again.  That should show better information.

Another thing to do is to run `solaar -dd` and try pressing some of the non-working keys.   That might show something about what the keys are actually doing.   You will have to have a window with the Solaar output open when you are pressing a key so that you can label the Solaar logging output corresponding to a key, if there is any.

**Comment by jontywigster** [comment]
I forgot to mention that I'd looked at solaar's output before, which wasn't added to when pressing a non-working key. Other keys did lead to output.  That was without -dd however. 


> I put together a PR #1175 

Many thanks indeed. I've used this and - 

>Another thing to do is to run solaar -dd and try pressing some of the non-working keys.

[solaar_dd_show_pr1175.txt](https://github.com/pwr-Solaar/Solaar/files/6487708/solaar_dd_show_pr1175.txt) -  contains `**good button**` before pressing a key that works; Text `**bad button**` is just before pressing a key that doesn't work.


> The thing I was looking for was the interaction between Solaar and the device when Solaar is determining which reprogrammable keys are available.

Thank you for this too. I was curious. Unfortunately I'm now more curious still. I hope it's OK to ask one more question. When I looked at usbmon, I saw traffic only for working remote control buttons. Does this, in itself, have any significance?

Many thanks again.

**Comment by pfps** [comment]
It is easier to see what is going on with the Solaar GUI.   Start Solaar as `bin/solaar -dd`.  Wait until the initial flood of messages ends.  Touch a key and see whether there is any logging information.   Do that for good keys and bad keys.  (The problem with `bin/solaar -dd show` is that it is hard to determine what communication is from Solaar querying the remote and what is from pressing keys on the remote.

My current feeling is that there are two possibilities.  The remote may be communicating with the hub using a completely different protocol, in which case there might be nothing that can be done.  The other possibility is that the remote is sending special notifications for some keys, which are ignored by Linux input drivers but that Solaar might be able to do something with, and setting up these notifications is done using the usual Logitech HID++ commands, which Solaar might be able to do.

The fact that you are seeing no traffic on usbmon is troubling, though.   It may be that some super-special setup is needed for the remote.

**Comment by pfps** [comment]
Also troubling is that no useful information about the 12 reprogrammable keys shows up with PR #1175.  I can see that the other feature is being interrogated but the results are all zeros.

**Comment by jontywigster** [comment]
Hello, 

> Start Solaar as bin/solaar -dd. Wait until the initial flood of messages ends. Touch a key and see whether there is any logging information. Do that for good keys and bad keys.

Sorry, I mentioned before that I'd originally seen no output when running without -dd. I meant without show. I've just double-checked I'm running and typing here the right thing, `bin/solaar -dd`. After the flood, the 'bad' keys indeed output nothing. Other keys seem fine. 

> My current feeling is that there are two possibilities. The remote may be communicating with the hub using a completely different protocol, in which case there might be nothing that can be done. The other possibility is that the remote is sending special notifications for some keys

Although only anecdotal, on a home automation forum I'm comparing notes with someone with the same model Unifying Receiver and remote control. They have missing keys in evtest also but, tantalisingly, missing for them is the opposite set of keys. I wonder if Logitech has a relevant mode or setting but, if there is, it's not obviously exposed. It just so happens that the missing keys aren't an issue for them so I wouldn't be able to provide any solaar output.

It feels like this issue is a bit unfair to be asking about. It's pretty niche!

**Comment by jontywigster** [comment]
Sorry about that, it was unintentionally incredibly rude. Ironically, as I mentioned, I was going to ask if I should close this ticket. I'm really grateful for the time you've spent so far.

**Comment by pfps** [comment]
It does look as if there is some hidden control that is used to modify how the 
remote works.   That probably means that nothing can be done without 
documentation.


One thing to try, if you have a Windows box, is to see whether you can get all 
the keys to work with Windows.  You will need to download and run Logitech 
Options and see whether that software can do some magic.


peter

**Comment by jontywigster** [comment]
Hello, 

Thanks for the additional tips. Windows itself recognises only the volume and mute buttons. Logitech Options states 'No devices detected' but did prompt to say my keyboard (which can only be the remote) would work best if I install Logitech SetPoint. SetPoint does recognise the remote as a keyboard but, sadly, offers no options.

I did try to look into how USB works, comparing notes with Wireshark traces. Unfortunately my confusion started early on when Windows and the Windows SDK agreed that my VM has three USB devices but the Logitech Unifying software sends a SET_REPORT request to a fourth, which doesn't exist but replies anyway.

I don't imagine I'll get a response but I've logged a ticket with Logitech.

Many thanks indeed for looking at this, I appreciate you replying at all to such a niche issue. This time I'll close the ticket intentionally.
